compiler/src/iree/compiler/Codegen/LLVMGPU/ROCDLBufferInstructionsOptimization.cpp

```diff
@@ -0,0 +1,131 @@
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_ROCDLBUFFERINSTRUCTIONSOPTIMIZATIONPASS
+#include "iree/compiler/Codegen/LLVMGPU/ROCDLPasses.h.inc"
+
+namespace {
+
+Value createi1And(Location loc, ArrayRef<Value> values, OpBuilder &builder) {
+  Value base =
+      builder.create<arith::IndexCastUIOp>(loc, builder.getI1Type(), values[0]);
+  for (auto value : values.drop_front()) {
+    Value rhs =
+        builder.create<arith::IndexCastUIOp>(loc, builder.getI1Type(), value);
+    base = builder.create<arith::AndIOp>(loc, base, rhs);
+  }
+  return base;
+}
+// Determine if the mask vector is all ones or all zeros and if so, then replace
+// the masked transferReadOp with transferReadOp with no mask for when the mask
+// is all ones and the padding values when the mask is all zeros. The pattern
+// thus does the following optimization.
+// clang-format off
+// mask = vector.create_mask %0, ..., %n, %c8 : vector<1x ... x1x8xi1>
+// %read = vector.transfer_read %memref, %mask : memref<..., amdgpu.raw_fat_buffer>
+// becomes
+// %padding = arith.constant dense<0> : vector<1x ... x1x8xbf16>
+// %read = vector.transfer_read %memref : memref<..., amdgpu.raw_fat_buffer> // no mask!
+// %masked_read
+//   = arith.select %0 && ... && %n ? %read : %padding : index,vector<1x ... x1x8xbf16>
+// clang-format on
+class SimplifyMaskVectorTransferRead
+    : public OpRewritePattern<vector::CreateMaskOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::CreateMaskOp maskOp,
+                                PatternRewriter &rewriter) const override {
+    auto loc = maskOp.getLoc();
+    auto readOp = dyn_cast<vector::TransferReadOp>(
+        *(maskOp.getResult().getUsers().begin()));
+    if (!readOp) {
+      return failure();
+    }
+
+    auto sourceType = dyn_cast<MemRefType>(readOp.getSource().getType());
+    if (!sourceType) {
+      return failure();
+    }
+    // This optimization only works on fat raw buffers.
+    if (!hasAMDGPUFatRawBufferAddressSpace(sourceType)) {
+      return failure();
+    }
+
+    SmallVector<bool> inBounds = readOp.getInBoundsValues();
+    if (inBounds.size() != sourceType.getRank()) {
+      return failure();
+    }
+
+    SmallVector<Value> maskIndices = maskOp.getOperands();
+    ArrayRef<int64_t> maskShape = maskOp.getResult().getType().getShape();
+    SmallVector<Value> ValuesToAnd;
+    for (auto [idx, i] : llvm::enumerate(maskIndices)) {
+      auto constantValue = getConstantIndex(i);
+      // We only support the pattern for reads that are in-bounds.
+      if (inBounds[idx] != true) {
+        return failure();
+      }
+      if (constantValue) {
+        // We bail-out if we can statically determine that a mask dim is not
+        // full.
+        if (maskShape[idx] != constantValue) {
```

Comment:
What I'll note here is that you need to be either all in or all out, so there might be ways to relax this condition?

Reply:
Hmmm, if you dont satisfy this condition I think you need to keep the mask as you only want some elements from the load so you cant be all in or all out?